### ISSUE TITLE
Add `validate` command

### ISSUE THREAD

**Issue by jqnatividad** [jqnatividad/qsv#46]
Checks a CSV against a jsonschema file.

The jsonschema file can be located on the filesystem or a URL.

**Comment by mhuang74** [comment]
Some good ideas for reference:

https://github.com/theodi/shared/issues/142

**Comment by mhuang74** [comment]
@jqnatividad Please let me know if you have some thoughts on the primary usecases.

* identify all schema/value errors in CSV and fix them
* quickly determine if CSV data matches expected schema
* generate schema from CSV

Here's 1st pass at docopt

```
Validate CSV data with JSON Schema.

Usage:
    qsv validate [options] [<input>] <json-schema>

fetch options:
    -c, --new-column <name>    Put error(s) in a new column instead.
    --fail-fast                Stops on first error.
    --generate-schema          Generate JSON Schema from data.


Common options:
    -h, --help                 Display this message
    -o, --output <file>        Write output to <file> instead of stdout.
    -n, --no-headers           When set, the first row will not be interpreted
                               as headers. Namely, it will be sorted with the rest
                               of the rows. Otherwise, the first row will always
                               appear as the header row in the output.
    -d, --delimiter <arg>      The field delimiter for reading CSV data.
                               Must be a single character. (default: ,)
    -q, --quiet                Don't show progress bars.
```

**Comment by jqnatividad** [comment]
As for the docopt format, it looks good, but we can tackle that once we agree on the general approach.

However, I think this is where multi-threading can really make a big difference, as we can break up the validation job as each row can be validated independently.

**Comment by mhuang74** [comment]
@jqnatividad here's my POC for `validate` command using jasonschema-rs crate.

https://github.com/jqnatividad/qsv/pull/145

Example run documented in [Validate.md](https://github.com/mhuang74/qsv/blob/validate/docs/Validate.md).

**Comment by jqnatividad** [comment]
@mhuang74 thanks for this.

The  problem is the progress bar does a count ala `qsv count`.  And since the file is invalid, it panics since the header doesn't match.

The fix is relatively simple, just set `flexible` to true setting up the Config ReaderBuilder.

And while we're on the progress bar, you may want to use the new  `update_cache_info!` macro.

I'll add comments to the PR for both changes...

**Comment by mhuang74** [comment]
thanks @jqnatividad. Above problem resolved. Simple csv check integration tests passing now.

I need to add another test to properly check for output files.

**Comment by mhuang74** [comment]
thanks @jqnatividad. Above problem resolved. Simple csv check integration tests passing now.

I need to add another test to properly check for output files.

Any suggestions on how to check for output files?

**Comment by jqnatividad** [comment]
Great! 

As for testing the output files, having a sample csv and corresponding, relatively complex jsonschema (the [PublicToilets](https://csvlint.io/schemas/53c7e25b63737603e3540300) schema on csvlint seems interesting and relatively complex), with expected valid and invalid rows should be good enough.

**Comment by mhuang74** [comment]
@jqnatividad  Need some help understanding standards around json schema, and whether choice of [jsonschema](https://crates.io/crates/jsonschema) crate is the right approach. Looks like it is working for dependent project [cjval](https://crates.io/crates/cjval) validating CityJSON files.

Looks like the [Public Toilets json schema](https://github.com/esd-org-uk/schemas/blob/master/PublicToilets/PublicToilets.json) conforms to [JSON Table Schema](http://dataprotocols.org/json-table-schema/) rather than the [JSONSchema ](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.1.1) which jsonschema-rs crate assumes. 

I attempted to run the public toilet schema through jsonschema-rs for validation, and it did not pick up and apply the constraints.

Specifically, the "keywordLocation" and "instanceLocation" are all blank.

what jsonschema-rs outputs with example People schema from JSON Schema site
```
[src/cmd/validate.rs:143] &validation_result = Object({
    "valid": Bool(
        false,
    ),
    "errors": Array([
        Object({
            "keywordLocation": String(
                "/properties/age/type",
            ),
            "instanceLocation": String(
                "/age",
            ),
            "absoluteKeywordLocation": String(
                "https://example.com/properties/age/type",
            ),
            "error": String(
                "\"21\" is not of type \"integer\"",
            ),
        }),
    ]),
})

```

snippet what jsonschema-rs outputs with Public Toilet schema

```
[src/cmd/validate.rs:143] &validation_result = Object({
    "valid": Bool(
        true,
    ),
    "annotations": Array([
        Object({
            "keywordLocation": String(
                "",
            ),
            "instanceLocation": String(
                "",
            ),
            "annotations": Object({
                "fields": Array([
                    Object({
                        "name": String(
                            "ExtractDate",
                        ),
                        "title": String(
                            "Extract Date",
                        ),
                        "description": String(
                            "The date that the data was last extracted from its source database or manually updated",
                        ),
                        "constraints": Object({
                            "required": String(
                                "true",
                            ),
                            "type": String(
                                "http://www.w3.org/2001/XMLSchema#string",
                            ),
                            "pattern": String(
                                "(^([0-2]\\d|[3][0-1])\\/([0]\\d|[1][0-2])\\/[1-2]\\d{3}((\\s([0-1]\\d|[2][0-3]):[0-5]\\d(:[0-5]\\d)?)?)|)$",
                            ),
                        }),
                    }),


```

**Comment by mhuang74** [comment]
Thanks @jqnatividad. They all require hand create schema file, so I stuck with public toilet dataset.

Integration test uses real data and illustrates invalid row showing up in output.invalid file.

**Comment by jqnatividad** [comment]
Public toilet it is! :)

Lemme know when you want me to merge the PR. Convert it to draft in the meantime?

**Comment by mhuang74** [comment]
Thanks @jqnatividad. I think it's ready for beta release. What do you think?

**Comment by jqnatividad** [comment]
Yep! Prepping for 0.3.0 release... :)

Did some minor tweaks though - https://github.com/jqnatividad/qsv/pull/148

**Comment by mhuang74** [comment]
Thanks for the fixups @jqnatividad . Good catch on progress bar. And just realized that named parameters are supported.

**Comment by jqnatividad** [comment]
@mhuang74, our MSRV is 1.58.1, so might as well use the latest features.

And I tweaked `validate` further just now to create the error-report as a jsonl file - easier to digest for both humans and machines :smirk:  with the ability to convert it to csv too with the `jsonl` command - https://github.com/jqnatividad/qsv/pull/149

**Comment by jqnatividad** [comment]
@mhuang74 and 0.3.0 featuring `validate` is released!

However, I had to limit cross-compilation for now, so binaries for certain platforms are not published (https://github.com/jqnatividad/qsv/commit/c54a2f27b888180cadb9648ae587d1c2965caa71).

Good news is that I managed to trace the problem to jsonschema, and the fix has been merged (https://github.com/Stranger6667/jsonschema-rs/pull/336), and we'll be able to publish binaries for those platforms again in the next release.

Thanks heaps again for your contributions!!!

Are you up to taking on `schema` (https://github.com/jqnatividad/qsv/issues/60)? :wink:  It need not use `stats` and `frequency` for the first release, it can just use [infers-jsonschema](https://github.com/Stranger6667/infers-jsonschema)...
